File: ros_ws/stm32_ws/src/camera_processing/src/distance.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <custom_msg/obj_msgs.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/opencv.hpp>
#include <opencv2/calib3d.hpp>
#include <cv_bridge/cv_bridge.h>
#include <darknet_ros_msgs/BoundingBoxes.h>

// Camera parameters
double fx = 894.6195678710938;
double fy = 894.6195678710938;
double cx = 635.2359008789062;
double cy = 367.5578308105469;
double baseline = 0.0055;
double depth_scale = 0.001;
double yaw_cam = 0.0;

double test_X= 0.0;
double test_Y= 0.0;
double test_Z= 0.0;
double test_distance =0.0;
unsigned short test_depth_value =0;

custom_msg::obj_msgs object_info;

cv::Mat depth_roi;
cv_bridge::CvImagePtr cv_ptr;

void depthImageCallback(const sensor_msgs::ImageConstPtr& depth_msg)
{
    try
    {
        cv_ptr = cv_bridge::toCvCopy(depth_msg, sensor_msgs::image_encodings::TYPE_16UC1);
    }
    catch (cv_bridge::Exception& e)
    {
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
    }
}

void bboxCallback(const darknet_ros_msgs::BoundingBoxes::ConstPtr& msg) {
    int num_boxes = msg->bounding_boxes.size();

    for (int i = 0; i < num_boxes; i++) {
        int xmin = msg->bounding_boxes[i].xmin;
        int ymin = msg->bounding_boxes[i].ymin;
        int xmax = msg->bounding_boxes[i].xmax;
        int ymax = msg->bounding_boxes[i].ymax;
	if (!cv_ptr) {
	    ROS_ERROR("cv_ptr is null!");
	    return;
	}
	
	depth_roi = cv_ptr->image;
	unsigned short depth_value = depth_roi.at<ushort>((ymin + ymax) / 2,(xmin + xmax) / 2);
    double distance = depth_value*depth_scale;

    //Test value
    test_distance = depth_value*depth_scale;
    
    double u = xmin + (xmax - xmin)/ 2;
    double v = ymin + (ymax - ymin)/ 2;
	    
    double x_norm = distance*((u - cx) / fx);
    double y_norm = distance*((v - cy) / fy);
    double z_norm = distance;

    test_X= x_norm;
    test_Y= y_norm;
    test_Z= z_norm;
    object_info.distance[i] = distance;
    object_info.x[i] = x_norm;
    test_depth_value = depth_value;
    }
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "distance_node");
    ros::NodeHandle nh;

    // Subscribe to depth map topic
    ros::Subscriber depth_sub = nh.subscribe("camera/depth/image_rect_raw", 100, depthImageCallback);
    ros::Subscriber bbox_sub = nh.subscribe("darknet_ros/bounding_boxes", 100, bboxCallback);

    ros::Publisher object_pub = nh.advertise<custom_msg::obj_msgs>("object", 1000);

    ros::Rate loop_rate(1);

    while (ros::ok())
    {
        ROS_INFO("Distance to object: %f",test_distance);
        ROS_INFO("Depth value: %d",test_depth_value);
        object_pub.publish(object_info);
        ros::spinOnce();
        loop_rate.sleep();
    }
    return 0;
}

File: ros_ws/stm32_ws/src/camera_processing/scripts/distance.py
import rospy
from sensor_msgs.msg import Image
from custom_msg.msg import obj_msgs
from sensor_msgs.msg import image_encodings
from cv_bridge import CvBridge
from darknet_ros_msgs.msg import BoundingBoxes
import numpy as np
import cv2


class DistanceNode:
    def __init__(self):
        # Camera parameters
        self.fx = 894.6195678710938
        self.fy = 894.6195678710938
        self.cx = 635.2359008789062
        self.cy = 367.5578308105469
        self.baseline = 0.0055
        self.depth_scale = 0.001
        self.yaw_cam = 0.0

        self.test_X = 0.0
        self.test_Y = 0.0
        self.test_Z = 0.0
        self.test_distance = 0.0
        self.test_depth_value = 0

        self.object_info = obj_msgs()
        self.object_info.distance = []
        self.object_info.x = []

        self.depth_roi = None
        self.cv_bridge = CvBridge()

        rospy.Subscriber("camera/depth/image_rect_raw", Image, self.depth_image_callback)
        rospy.Subscriber("darknet_ros/bounding_boxes", BoundingBoxes, self.bbox_callback)

        self.object_pub = rospy.Publisher("object", obj_msgs, queue_size=1000)

        self.rate = rospy.Rate(1)  # 1 Hz

    def depth_image_callback(self, depth_msg):
        try:
            self.depth_roi = self.cv_bridge.imgmsg_to_cv2(depth_msg, desired_encoding="passthrough")
        except CvBridgeError as e:
            rospy.logerr("CvBridge exception: %s", e)

    def bbox_callback(self, msg):
        num_boxes = len(msg.bounding_boxes)
        self.object_info.distance.clear()
        self.object_info.x.clear()

        for i in range(num_boxes):
            xmin = msg.bounding_boxes[i].xmin
            ymin = msg.bounding_boxes[i].ymin
            xmax = msg.bounding_boxes[i].xmax
            ymax = msg.bounding_boxes[i].ymax

            if self.depth_roi is None:
                rospy.logerr("depth_roi is None!")
                return

            depth_value = self.depth_roi[int((ymin + ymax) / 2), int((xmin + xmax) / 2)]
            distance = depth_value * self.depth_scale

            # Test value
            self.test_distance = distance

            u = xmin + (xmax - xmin) / 2
            v = ymin + (ymax - ymin) / 2

            x_norm = distance * ((u - self.cx) / self.fx)
            y_norm = distance * ((v - self.cy) / self.fy)
            z_norm = distance

            self.test_X = x_norm
            self.test_Y = y_norm
            self.test_Z = z_norm

            self.object_info.distance.append(distance)
            self.object_info.x.append(x_norm)

            self.test_depth_value = depth_value

    def run(self):
        rospy.init_node("distance_node")
        rospy.loginfo("Starting distance_node")

        while not rospy.is_shutdown():
            rospy.loginfo("Distance to object: %f", self.test_distance)
            rospy.loginfo("X: %f", self.test_X)
            self.object_pub.publish(self.object_info)
            self.rate.sleep()


if __name__ == "__main__":
    try:
    	distance_node = DistanceNode()
    	distance_node.run()
    except rospy.ROSInterruptException:
        pass
